### ISSUE TITLE
"Create before Delete" rather than "Delete before Create" by default

### ISSUE THREAD

**Issue by scottyw** [lyraproj/lyra#160]
When an immutable attribute changes, Lyra needs to delete the old version of the resource and create a new one. Today we delete first then recreate once deletion has succeeded. That causes problems when resources have dependent resources that prevent their deletion.

For example, start with a workflow like this:
```yaml
aws_vpc_instance_tf_yaml:
  typespace: TerraformAws
  activities:
    aws_vpc:
      output: aws_vpc_id
      state:
        cidr_block: 192.168.0.0/20
        enable_classiclink_dns_support: false
    aws_subnet:
      state:
        vpc_id: $aws_vpc_id
        cidr_block: 192.168.1.0/24
```

Changing a mutable field on the VPC (like `enable_classiclink_dns_support`) works fine because it is modified in place:
```yaml
aws_vpc_instance_tf_yaml:
  typespace: TerraformAws
  activities:
    aws_vpc:
      output: aws_vpc_id
      state:
        cidr_block: 192.168.0.0/20
        enable_classiclink_dns_support: true
    aws_subnet:
      state:
        vpc_id: $aws_vpc_id
        cidr_block: 192.168.1.0/24
```

Changing an immutable field on the VPC (like `cidr_block`) forces Lyra to delete and recreate the resource.
```yaml
aws_vpc_instance_tf_yaml:
  typespace: TerraformAws
  activities:
    aws_vpc:
      output: aws_vpc_id
      state:
        cidr_block: 192.168.0.0/16
        enable_classiclink_dns_support: false
    aws_subnet:
      state:
        vpc_id: $aws_vpc_id
        cidr_block: 192.168.1.0/24
```

Deletion fails in this case because the existence of the subnet causes a dependency violation:
```
panic: invocation of TerraformAws::Aws_vpcHandler delete failed: EVAL_GO_FUNCTION_ERROR Go function Delete returned error 'DependencyViolation: The vpc 'vpc-0f9e4ec737076b29e' has dependencies and cannot be deleted.
	status code: 400, request id: 588fad00-7183-4dea-b7fa-bfa24d79bc38'
```

@markfuller, @thallgren and I discussed this recently and concluded we might address this sort of problem by creating the new resource first and allowing garbage collection to delete any leftover resources at the end of the run.

This has several advantages. Primarily, we avoid the dependency violation! Beyond that, eager deletion of a resource would result in the complete deletion of all ancestor resources e.g. the VPC, all subnets, children of subnets, grandchildren of subnets etc. In some cases that might be necessary but in some it will be possible to reparent some of those ancestor resources to the newly created parents. For example, it is in principle possible that we'd have to delete the VPC but could reparent the existing subnets to the new VPC rather than forcing their recreation too. (In practice that's not the case for subnets but illustrates the idea.)

There are also disadvantages, for example, if there is a limit on the number of instances of a resource that can exist (say, no more than 5 VPCs total, or no more than one S3 bucket with a given name). We're proposing adding support for a metaparameter called "eager delete" that would allow the user to indicate in their workflow (or some such) that specific resources should be deleted eagerly. We already have a mechanism to allow users to express relationships between resources in a way that would allow Lyra to delete dependent resource first.

In summary, we need to choose a default policy for cases where we have no special information about what to do. Moving from "delete first" to "recreate first" seems on balance to have advantages.
